stream_alert/alert_processor/outputs/output_base.py

```diff
@@ -243,12 +250,90 @@ def _kms_decrypt(self, data):
             str: Decrypted json string
         """
         try:
-            client = boto3.client('kms', region_name=self.region)
+            client = boto3.client('kms', region_name=self._region)
             response = client.decrypt(CiphertextBlob=data)
             return response['Plaintext']
         except ClientError as err:
             LOGGER.error('an error occurred during credentials decryption: %s', err.response)
 
+    @staticmethod
+    def get_formatted_output_credentials_name(service_name, descriptor):
+        """Formats the output name for this credential by combining the service
+        and the descriptor.
+
+        Args:
+            service_name (str): Service name on output class (i.e. "pagerduty", "demisto")
+            descriptor (str): Service destination (ie: slack channel, pd integration)
+
+        Returns:
+            str: Formatted credential name (ie: slack_ryandchannel)
+        """
+        cred_name = str(service_name)
+
+        # should descriptor be enforced in all rules?
+        if descriptor:
+            cred_name = '{}/{}'.format(cred_name, descriptor)
+
+        return cred_name
+
+    def get_aws_account_id(self):
+        """Returns the AWS account ID"""
+        return self._account_id
+
+    def get_secrets_bucket_name(self):
+        """Returns the bucket name of the S3 bucket to look in for encrypted credentials"""
+        return self._secrets_bucket
+
+
+class OutputDispatcher(object):
+    """OutputDispatcher is the base class to handle routing alerts to outputs
+
+    Public methods:
+        get_secrets_bucket_name: returns the name of the s3 bucket for secrets that
+            includes a unique prefix
+        get_config_service: the name of the service used by the config to store any
+            configured outputs for this service. implemented by some subclasses, but
+            subclass is not required to implement
+        format_output_config: returns a formatted version of the outputs configuration
+            that is to be written to disk
+        get_user_defined_properties: returns any properties for this output that must be
+            provided by the user. must be implemented by subclasses
+        dispatch: handles the actual sending of alerts to the configured service. must
+            be implemented by subclass
+    """
+    __metaclass__ = ABCMeta
+    __service__ = NotImplemented
+
+    # How many times it will attempt to retry something failing using backoff
+    MAX_RETRY_ATTEMPTS = 5
+
+    # _DEFAULT_REQUEST_TIMEOUT indicates how long the requests library will wait before timing
+    # out for both get and post requests. This applies to both connection and read timeouts
+    _DEFAULT_REQUEST_TIMEOUT = 3.05
+
+    def __init__(self, config):
+        self.region = REGION
+        self.config = config
+
+        self._credentials_provider = OutputCredentialsProvider(config,
```

Comment:
I know this format is _technically_ valid (and likely not even your doing) but can we do this instead:

```
...
    self._credentials_provider = OutputCredentialsProvider(
        config,
        self._get_default_properties(),
        self.__service__
    )
...
```